manim/mobject/mobject.py

```diff
@@ -1731,6 +1731,29 @@ def get_boundary_point(self, direction):
         index = np.argmax(np.dot(all_points, np.array(direction).T))
         return all_points[index]
 
+    def get_midpoint(self) -> np.ndarray:
```

Comment:
I think there should be multiple implementations of `get_midpoint`. For instance, what if the user wanted to get the midpoint between self and another Mobject/point? In which case there should be an overload for this.

Reply:
Additionally, this should not be implemented in Mobject since Mobject itself doesn't actually implement `point_from_proportion`. For instance, how would this apply to ImageMobject? ImageMobject would receive access to this method (being a Mobject), but it doesn't overload `point_from_proportion` so this wouldn't work for all Mobject. A better place to place this would be VMobject.

Reply:
`Mobject` doesn't implement `point_from_proportion` but defines it with the expectation that child classes will implement it, essentially an abstract method. And so, I think it's totally valid for things defined under `Mobject` to use `point_from_proportion` because it's supposed to exist, and if it doesn't then you can't use the default implementation of `get_midpoint`, but it can be overridden just fine, or if not then the user has violated a contract of sorts by using `get_midpoint` on a mobject that doesn't define `point_from_proportion`. Particularly I don't think this belongs under `VMobject` because more mobject types than `VMobject` can have `point_from_proportion` implementations

Reply:
Sure, I forget about the odd case of PMobject (which frankly, I've never used/seen used). I still think this should have an overload though before merge since `Mobject.get_midpoint(other_Mobject)` is actually applicable to all Mobjects.

Reply:
@friedkeenan makes a valid point, but I have a bad feeling that this will make users think `get_midpoint()` is usable on any Mobject.

Reply:
This is already the case for other mobject methods, such as `pointwise_become_partial` and `interpolate_color` which are not implemented by `Mobject` but are still used in other methods. I don't think it's much of a problem to do the same for `point_from_proportion`.

Reply:
I don't think that just because something is already the case is a good justification for continuing the practice.

Reply:
If the user uses the method on something that doesn't have `point_from_proportion` implemented, then they'll get a `NotImplementedError` thrown in `Mobject.point_from_proportion`. I think that's clear enough to the user that `get_midpoint` requires a mobject that implements `point_from_proportion`.

Reply:
I'm not a huge fan of that error for end users because while it does make clear there _was_ an error, it does not make clear to end users _why_ it's not implemented. Is it something they did wrong? (From their perspective) Is it just because manim is far behind in development that this feature isn't done yet? I think we should have clearer messages behind the `NotImplementedError`. (I am aware that it says "Please override in a child class")

Reply:
I think the message could be clearer, yeah, and perhaps it could be marked with `abc.abstractmethod` to make sure child classes implement it (though I suppose that would cause issues with instantiating `Mobject` directly, along with issues that would come from `ValueTracker`), but I don't see that as a problem with the implementation of `get_midpoint` being under `Mobject`.

Reply:
As long as we provide a clear error message when the method is being called for a Mobject, I am fine with it.

Reply:
What if we implement point_from_proportion and get_midpoint only in the child classes and remove it completely from Mobject?